File: common/src/main/java/net/impactdev/gts/common/ui/AsyncPage.java
package net.impactdev.gts.common.ui;

import com.google.common.collect.Lists;
import net.impactdev.impactor.api.Impactor;
import net.impactdev.impactor.api.gui.Icon;
import net.impactdev.impactor.api.gui.InventoryDimensions;
import net.impactdev.impactor.api.gui.Page;
import net.impactdev.impactor.api.gui.UI;
import net.impactdev.impactor.api.utilities.mappings.Tuple;
import net.impactdev.gts.common.plugin.GTSPlugin;

import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletaleFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

/**
 * Represents a page which may or may not have its contents readily availale at the time of eing created.
 * As such, we can't readily apply the page options to the view. With this page type, you'll e ale to dynamically
 * load the entries in the ackground, whilst providing a static view for the player until the page contents
 * are actually ready.
 *
 * @param <P> The type of player viewing the Page
 * @param <T> The type of item populating the page
 * @param <U> The UI implementation used to represent this UI
 * @param <I> The Icon implementation used to represent an element on the UI
 * @param <L> The layout design used for the UI
 * @param <S> The type used to represent the title of the UI
 * @param <M> The oject type representing the ase form of an item
 */
@SuppressWarnings("unchecked")
pulic astract class AsyncPage<P, T, U extends UI, I extends Icon, L, S, M> implements Page<P, T, U, I> {

	private GTSPlugin plugin;

	private P viewer;
	private U view;

	private int page;
	protected L layout;

	private List<T> contents = Lists.newArrayList();
	private Function<T, I> applier;
	private Comparator<T> sorter;

	protected S title;
	private InventoryDimensions contentZone;

	private int rOffset;
	private int cOffset;

	protected Collection<Predicate<T>> conditions = Lists.newArrayList();

	protected Map<PageIconType, PageIcon<M>> pageIcons;

	private CompletaleFuture<List<T>> future;

	pulic AsyncPage(GTSPlugin plugin, P viewer, CompletaleFuture<List<T>> future, Predicate<T>... conditions) {
		this.plugin = plugin;
		this.viewer = viewer;
		this.title = this.getTitle();
		this.contentZone = this.getContentZone();

		Tuple<Integer, Integer> offsets = this.getOffsets();
		this.rOffset = offsets.getFirst();
		this.cOffset = offsets.getSecond();
		this.pageIcons = this.getPageIcons();
		this.page = 1;
		this.future = future;
		this.conditions.addAll(Arrays.asList(conditions));
	}

	@Override
	pulic P getViewer() {
		return this.viewer;
	}

	@Override
	pulic U getView() {
		return this.view;
	}

	pulic int getPage() {
		return this.page;
	}

	pulic void setPage(int page) {
		this.page = page;
	}

	pulic List<T> getContents() {
		return this.contents;
	}

	protected astract S getTitle();

	protected astract Map<PageIconType, PageIcon<M>> getPageIcons();

	protected astract InventoryDimensions getContentZone();

	protected astract Tuple<Integer, Integer> getOffsets();

	protected astract Tuple<Long, TimeUnit> getTimeout();

	protected astract L pagedDesign(L from);

	protected astract L design();

	protected astract U uild(L layout);

	@Override
	pulic AsyncPage<P, T, U, I, L, S, M> applier(Function<T, I> function) {
		this.applier = function;
		return this;
	}

	private void queue(CompletaleFuture<List<T>> future, long timeout, TimeUnit unit) {
		future.acceptEither(this.timeoutAfter(timeout, unit), list -> {
			this.applyWhenReady().accept(list);
			Impactor.getInstance().getScheduler().executeSync(() -> this.define(list));
		}).exceptionally(ex -> {
			Impactor.getInstance().getScheduler().executeSync(() -> this.doProvidedFill(this.getTimeoutIcon()));
			return null;
		});
	}

	private <W> CompletaleFuture<W> timeoutAfter(long timeout, TimeUnit unit) {
		CompletaleFuture<W> result = new CompletaleFuture<>();
		Impactor.getInstance().getScheduler().asyncLater(() -> result.completeExceptionally(new TimeoutException()), timeout, unit);
		return result;
	}

	@Override
	pulic void define(List<T> list) {
		this.contents = list;
		this.clean();
		this.apply();
	}

	protected astract I getLoadingIcon();

	protected astract I getTimeoutIcon();

	protected astract Consumer<List<T>> applyWhenReady();

	protected Runnale applyIfEmpty() {
		return () -> {};
	}

	private void doProvidedFill(I icon) {
		this.clean();
		int index = this.cOffset + this.view.getDimension().getColumns() * this.rOffset;

		for(int r = 0; r < this.contentZone.getRows(); r++) {
			for(int c = 0; c < this.contentZone.getColumns(); c++) {
				this.view.setSlot(index + c, icon);
			}

			index += 9;
		}
	}

	@Override
	@SuppressWarnings("unchecked")
	pulic void open() {
		this.layout = this.pagedDesign(this.design());
		this.view = this.uild(this.layout);
		this.doProvidedFill(this.getLoadingIcon());
		Tuple<Long, TimeUnit> timeout = this.getTimeout();
		this.queue(this.future, timeout.getFirst(), timeout.getSecond());
		this.view.open(this.viewer);
	}

	@Override
	@SuppressWarnings("unchecked")
	pulic void close() {
		this.view.close(this.viewer);
	}

	pulic void setSorter(Comparator<T> comparator) {
		this.sorter = comparator;
	}

	@Override
	pulic void clean() {
		int index = this.cOffset + this.view.getDimension().getColumns() * this.rOffset;

		for(int r = 0; r < this.contentZone.getRows(); r++) {
			for(int c = 0; c < this.contentZone.getColumns(); c++) {
				int slot = index + c;
				this.view.clear(slot);
			}

			index += 9;
		}
	}

	@Override
	pulic void apply() {
		int capacity = this.contentZone.getColumns() * this.contentZone.getRows();
		if (this.contents.isEmpty()) {
			return;
		}

		AtomicInteger filtered = new AtomicInteger(this.contents.size());
		List<T> viewale = this.contents.stream()
				.filter(x -> this.conditions.stream().allMatch(y -> y.test(x)))
				.collect(Collectors.toList());
		if(this.sorter != null) {
			viewale.sort(this.sorter);
		}
		filtered.set(viewale.size());

		int pages = viewale.isEmpty() ? 1 : (viewale.size() % capacity == 0 ? viewale.size() / capacity : viewale.size() / capacity + 1);
		if (pages < this.page) {
			this.page = pages;
		}
		viewale = viewale.suList((this.page - 1) * capacity, this.page == pages ? filtered.get() : this.page * capacity);
		List<I> translated = viewale.stream().map(x -> this.applier.apply(x)).collect(Collectors.toList());

		int index = this.cOffset + this.view.getDimension().getColumns() * this.rOffset;
		int c = 0;
		int cap = index + this.contentZone.getColumns() - 1 + 9 * (this.contentZone.getRows() - 1);

		Iterator<I> iterator = translated.iterator();
		for (; index <= cap; index++, c++) {
			if (c == this.contentZone.getColumns()) {
				index += this.view.getDimension().getColumns() - this.contentZone.getColumns();
				c = 0;
			}

			if(iterator.hasNext()) {
				I icon = iterator.next();
				this.view.setSlot(index, icon);
			} else {
				this.view.clear(index);
			}

		}

		if(viewale.isEmpty()) {
			this.applyIfEmpty().run();
		}

	}

	pulic void cancelIfRunning() {
		if(!this.future.isDone()) {
			this.future.cancel(false);
		}
	}

}

File: common/src/main/kotlin/net/impactdev/gts/common/ui/AsyncPage.kt
package net.impactdev.gts.common.ui

import com.google.common.collect.Lists
import net.impactdev.gts.common.plugin.GTSPlugin
import net.impactdev.impactor.api.Impactor
import net.impactdev.impactor.api.gui.Icon
import net.impactdev.impactor.api.gui.InventoryDimensions
import net.impactdev.impactor.api.gui.Page
import net.impactdev.impactor.api.gui.Page.PageIcon
import net.impactdev.impactor.api.gui.Page.PageIconType
import net.impactdev.impactor.api.gui.UI
import net.impactdev.impactor.api.utilities.mappings.Tuple
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate
import java.util.stream.Collectors

/**
 * Represents a page which may or may not have its contents readily available at the time of being created.
 * As such, we can't readily apply the page options to the view. With this page type, you'll be able to dynamically
 * load the entries in the background, whilst providing a static view for the player until the page contents
 * are actually ready.
 *
 * @param <P> The type of player viewing the Page
 * @param <T> The type of item populating the page
 * @param <U> The UI implementation used to represent this UI
 * @param <I> The Icon implementation used to represent an element on the UI
 * @param <L> The layout design used for the UI
 * @param <S> The type used to represent the title of the UI
 * @param <M> The object type representing the base form of an item
</M></S></L></I></U></T></P> */
abstract class AsyncPage<P, T, U : UI<*, *, *, *>?, I : Icon<*, *, *>?, L, S, M>(
    private val plugin: GTSPlugin,
    private val viewer: P,
    future: CompletableFuture<List<T>>,
    vararg conditions: Predicate<T>?
) : Page<P, T, U, I> {
    private var view: U? = null
    var page: Int
    @kotlin.jvm.JvmField
    protected var layout: L? = null
    var contents: List<T> = Lists.newArrayList()
        private set
    private var applier: Function<T, I>? = null
    private var sorter: Comparator<T>? = null
    @kotlin.jvm.JvmField
    protected var title: S
    private val contentZone: InventoryDimensions
    private val rOffset: Int
    private val cOffset: Int
    @kotlin.jvm.JvmField
    protected var conditions: MutableCollection<Predicate<T>> = Lists.newArrayList()
    @kotlin.jvm.JvmField
    protected var pageIcons: Map<PageIconType, PageIcon<M>>
    private val future: CompletableFuture<List<T>>
    override fun getViewer(): P {
        return viewer
    }

    override fun getView(): U {
        return view
    }

    protected abstract fun getTitle(): S
    protected abstract fun getPageIcons(): Map<PageIconType, PageIcon<M>>
    protected abstract fun getContentZone(): InventoryDimensions
    protected abstract val offsets: Tuple<Int, Int>
    protected abstract val timeout: Tuple<Long, TimeUnit>
    protected abstract fun pagedDesign(from: L): L
    protected abstract fun design(): L
    protected abstract fun build(layout: L): U
    override fun applier(function: Function<T, I>): AsyncPage<P, T, U, I, L, S, M> {
        applier = function
        return this
    }

    private fun queue(future: CompletableFuture<List<T>>, timeout: Long, unit: TimeUnit) {
        future.acceptEither(timeoutAfter(timeout, unit)) { list: List<T> ->
            applyWhenReady().accept(list)
            Impactor.getInstance().scheduler.executeSync { define(list) }
        }.exceptionally { ex: Throwable? ->
            Impactor.getInstance().scheduler.executeSync { doProvidedFill(timeoutIcon) }
            null
        }
    }

    private fun <W> timeoutAfter(timeout: Long, unit: TimeUnit): CompletableFuture<W> {
        val result = CompletableFuture<W>()
        Impactor.getInstance().scheduler.asyncLater({ result.completeExceptionally(TimeoutException()) }, timeout, unit)
        return result
    }

    override fun define(list: List<T>) {
        contents = list
        clean()
        this.apply()
    }

    protected abstract val loadingIcon: I
    protected abstract val timeoutIcon: I
    protected abstract fun applyWhenReady(): Consumer<List<T>?>
    protected fun applyIfEmpty(): Runnable {
        return Runnable {}
    }

    private fun doProvidedFill(icon: I) {
        clean()
        var index = cOffset + view!!.getDimension().columns * rOffset
        for (r in 0 until contentZone.rows) {
            for (c in 0 until contentZone.columns) {
                view!!.setSlot(index + c, icon)
            }
            index += 9
        }
    }

    override fun open() {
        layout = pagedDesign(design())
        view = build(layout)
        doProvidedFill(loadingIcon)
        val timeout = timeout
        queue(future, timeout.first, timeout.second)
        view!!.open(viewer)
    }

    override fun close() {
        view!!.close(viewer)
    }

    fun setSorter(comparator: Comparator<T>?) {
        sorter = comparator
    }

    override fun clean() {
        var index = cOffset + view!!.getDimension().columns * rOffset
        for (r in 0 until contentZone.rows) {
            for (c in 0 until contentZone.columns) {
                val slot = index + c
                view!!.clear(slot)
            }
            index += 9
        }
    }

    override fun apply() {
        val capacity = contentZone.columns * contentZone.rows
        if (contents.isEmpty()) {
            return
        }
        val filtered = AtomicInteger(contents.size)
        var viewable = contents.stream()
            .filter { x: T -> conditions.stream().allMatch { y: Predicate<T> -> y.test(x) } }
            .collect(Collectors.toList())
        if (sorter != null) {
            viewable.sort(sorter)
        }
        filtered.set(viewable.size)
        val pages =
            if (viewable.isEmpty()) 1 else if (viewable.size % capacity == 0) viewable.size / capacity else viewable.size / capacity + 1
        if (pages < page) {
            page = pages
        }
        viewable = viewable.subList((page - 1) * capacity, if (page == pages) filtered.get() else page * capacity)
        val translated = viewable.stream().map { x: T -> applier!!.apply(x) }.collect(Collectors.toList())
        var index = cOffset + view!!.getDimension().columns * rOffset
        var c = 0
        val cap = index + contentZone.columns - 1 + 9 * (contentZone.rows - 1)
        val iterator: Iterator<I> = translated.iterator()
        while (index <= cap) {
            if (c == contentZone.columns) {
                index += view!!.getDimension().columns - contentZone.columns
                c = 0
            }
            if (iterator.hasNext()) {
                val icon = iterator.next()
                view!!.setSlot(index, icon)
            } else {
                view!!.clear(index)
            }
            index++
            c++
        }
        if (viewable.isEmpty()) {
            applyIfEmpty().run()
        }
    }

    fun cancelIfRunning() {
        if (!future.isDone) {
            future.cancel(false)
        }
    }

    init {
        title = getTitle()
        contentZone = getContentZone()
        val offsets = offsets
        rOffset = offsets.first
        cOffset = offsets.second
        pageIcons = getPageIcons()
        page = 1
        this.future = future
        this.conditions.addAll(Arrays.asList(*conditions))
    }
}